### DISCUSSION TITLE
Coupling Gadget-2 With ph4 and Multiples Using Bridge

### DISCUSSION THREAD

**Discussion by jeremykaram** [amusecode/amuse#976] [Q&A] [answered]
Hi all,
I am aiming to evolve a cluster of stars and gas with binary stars. I have bridged things together in the following way:
```

self.star_code = new_star_code_ph4()
self.gas_code = new_gas_code_gadget()

def new_code_BHTree():
             return BHTree(converter)

self.calculate_gravity_code_for_gas = bridge.CalculateFieldsForCodes(
             new_code_BHTree, input_codes=[self.gas_code]
)

self.multiples_code = multiples.Multiples(self.star_code, new_smalln_code, kepler_code, constants.G)

self.gas_to_star_codes = [self.calculate_gravity_code_for_gas]
self.star_to_gas_codes = [self.multiples_code]

self.bridge_system = bridge.Bridge(
             timestep=interaction_timestep, use_threading=False
)
self.bridge_system.add_system(
             self.gas_code,
             self.star_to_gas_codes
)
self.bridge_system.add_system(
             self.multiples_code,
             self.gas_to_star_codes
)

self.code = self.bridge_system
```
My question is the following: should I be bridging Gadget-2 to multiples in this way? Or, should I switch the multiples_code in Bridge to my ph4 code that I initiated before? i.e. should ph4 be my star_to_gas_codes instead of multiples_code?

I am quite new to multiples so I guess another way to ask this question is: does multiples return the evolution of every star in a cluster (including those not detected as an encounter) or does it just return the stars determined to be an encounter by multiples and send those to ph4 to evolve the whole system?

Please let me know if you need more detail. 
Thank you,
Jeremy Karam

**Comment by rieder** [comment, marked as answer]
Hi Jeremy,

Thanks for asking - this is something we hope to expand on in the second edition of the AMUSE book.
I'll explain here briefly what exactly happens.
First, you set up two integrators: Multiples for the stars, and Gadget for the gas.
Second, you initiate a Bridge system (the parameters don't matter for this explanation).
Third, you add your systems to Bridge. The arguments here are the following: 
1. the system being added,
2. a list or tuple of everything kicking the system

In this case, you want the stars in Multiples to be kicked by the gas in Gadget, and vice versa.
What you specify here is /how/ those kicks (from the particles in 2 to the particles in 1) should be calculated.

Since Gadget is a tree code, it makes sense to treat any forces involving it with a tree code as well - no need for anything more fancy.
This goes for both gas-to-stars and stars-to-gas!

So, to answer your question, I would probably not use multiples or ph4 for stars_to_gas, but instead use the same approach as for gas_to_stars, with a `new_code_BHTree` (or another tree code to your liking - BHTree isn't the fastest available).

Hope this helps!
Steven

**Reply by jeremykaram** [reply]
Hi Steven,
Thank you for the response. This has definitely cleared a few things up. I will play around a little more and post again if I come across any brick walls. Thanks again!
